File: code/go/ch09/src/luago/state/lua_stack.go
package state

type luaStack struct {
	/* virtual stack */
	slots []luaValue
	top   int
	/* call info */
	closure *closure
	varargs []luaValue
	pc      int
	/* linked list */
	prev *luaStack
}

func newLuaStack(size int) *luaStack {
	return &luaStack{
		slots: make([]luaValue, size),
		top:   0,
	}
}

func (self *luaStack) check(n int) {
	free := len(self.slots) - self.top
	for i := free; i < n; i++ {
		self.slots = append(self.slots, nil)
	}
}

func (self *luaStack) push(val luaValue) {
	if self.top == len(self.slots) {
		panic("stack overflow!")
	}
	self.slots[self.top] = val
	self.top++
}

func (self *luaStack) pop() luaValue {
	if self.top < 1 {
		panic("stack underflow!")
	}
	self.top--
	val := self.slots[self.top]
	self.slots[self.top] = nil
	return val
}

func (self *luaStack) pushN(vals []luaValue, n int) {
	nVals := len(vals)
	if n < 0 {
		n = nVals
	}

	for i := 0; i < n; i++ {
		if i < nVals {
			self.push(vals[i])
		} else {
			self.push(nil)
		}
	}
}

func (self *luaStack) popN(n int) []luaValue {
	vals := make([]luaValue, n)
	for i := n - 1; i >= 0; i-- {
		vals[i] = self.pop()
	}
	return vals
}

func (self *luaStack) absIndex(idx int) int {
	if idx >= 0 {
		return idx
	}
	return idx + self.top + 1
}

func (self *luaStack) isValid(idx int) bool {
	absIdx := self.absIndex(idx)
	return absIdx > 0 && absIdx <= self.top
}

func (self *luaStack) get(idx int) luaValue {
	absIdx := self.absIndex(idx)
	if absIdx > 0 && absIdx <= self.top {
		return self.slots[absIdx-1]
	}
	return nil
}

func (self *luaStack) set(idx int, val luaValue) {
	absIdx := self.absIndex(idx)
	if absIdx > 0 && absIdx <= self.top {
		self.slots[absIdx-1] = val
		return
	}
	panic("invalid index!")
}

func (self *luaStack) reverse(from, to int) {
	slots := self.slots
	for from < to {
		slots[from], slots[to] = slots[to], slots[from]
		from++
		to--
	}
}

File: code/csharp/ch10/luavm/state/LuaStack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Threading;
using luavm.api;

namespace luavm.state
{
    public class LuaStack
    {
        public object[] slots;
        internal int top;
        internal LuaStack prev;
        internal Closure closure;
        internal object[] varargs;
        internal int pc;
        private LuaState state;
        internal Dictionary<int, Upvalue> openuvs;

        internal static LuaStack newLuaStack(int size, LuaState state)
        {
            return new LuaStack
            {
                slots = new object[size],
                top = 0,
                state = state
            };
        }

        internal void check(int n)
        {
            var free = slots.Length - top;
            var slotList = slots.ToList();
            for (var i = free; i < n; i++)
            {
                slotList.Add(null);
            }

            slots = slotList.ToArray();
        }

        internal void push(object val)
        {
            if (top == slots.Length)
            {
                throw new Exception("stack overflow!");
            }

            slots[top] = val;

            top++;
        }

        internal int absIndex(int idx)
        {
            if (idx >= 0 || idx < Consts.LUA_REGISTRYINDEX)
            {
                return idx;
            }

            return idx + top + 1;
        }

        internal bool isValid(int idx)
        {
            if (idx < Consts.LUA_REGISTRYINDEX)
            {
                var uvIdx = Consts.LUA_REGISTRYINDEX - idx - 1;
                var c = closure;
                return c != null && (uvIdx < c.upvals.Length);
            }

            var absIdx = absIndex(idx);
            return absIdx > 0 && absIdx <= top;
        }

        internal object get(int idx)
        {
            if (idx < Consts.LUA_REGISTRYINDEX)
            {
                var uvIdx = Consts.LUA_REGISTRYINDEX - idx - 1;
                var c = closure;
                if (c == null || uvIdx >= c.upvals.Length)
                {
                    return null;
                }

                return c.upvals[uvIdx].val;
            }

            if (idx == Consts.LUA_REGISTRYINDEX)
            {
                return state.registry;
            }

            var absIdx = absIndex(idx);
            if (absIdx > 0 && absIdx <= top)
            {
                return slots[absIdx - 1];
            }

            return null;
        }

        internal object pop()
        {
            if (top < 1)
            {
                throw new Exception("stack overflow!");
            }

            top--;
            var val = slots[top];
            slots[top] = null;
            return val;
        }

        internal void pushN(object[] vals, int n)
        {
            var nVals = vals.Length;
            if (n < 0)
            {
                n = nVals;
            }

            for (var i = 0; i < n; i++)
            {
                if (i < nVals)
                {
                    push(vals[i]);
                }
                else
                {
                    push(null);
                }
            }
        }

        internal object[] popN(int n)
        {
            var vals = new object[n];
            for (var i = n - 1; i >= 0; i--)
            {
                vals[i] = pop();
            }

            return vals;
        }


        internal void set(int idx, object val)
        {
            if (idx < Consts.LUA_REGISTRYINDEX)
            {
                var uvIdx = Consts.LUA_REGISTRYINDEX - idx - 1;
                var c = closure;
                if (c != null && uvIdx < c.upvals.Length)
                {
                    c.upvals[uvIdx].val = val;
                    return;
                }
            }

            var absIdx = absIndex(idx);
            if (absIdx <= 0 || absIdx > top)
            {
                throw new Exception("invalid index!");
            }

            slots[absIdx - 1] = val;
        }

        internal void reverse(int from, int to)
        {
            if (to > from)
            {
                Array.Reverse(slots, from, to - from + 1);
            }
            else if (to < from)
            {
                Array.Reverse(slots, to, from - to + 1);
            }
        }
    }
}